File: software/examples/c/example_callback.c
#include <stdio.h>

// FIXME: This example is incomplete

#include "ip_connection.h"
#include "bricklet_rgb_led_button.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change XYZ to the UID of your RGB LED Button Bricklet

// Callback function for button state changed callback
void cb_button_state_changed(uint8_t state, void *user_data) {
	(void)user_data; // avoid unused parameter warning

	printf("State: %d\n", state);
}

int main(void) {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	RGBLEDButton rlb;
	rgb_led_button_create(&rlb, UID, &ipcon);

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		return 1;
	}
	// Don't use device before ipcon is connected

	// Register button state changed callback to function cb_button_state_changed
	rgb_led_button_register_callback(&rlb,
	                                 RGB_LED_BUTTON_CALLBACK_BUTTON_STATE_CHANGED,
	                                 (void *)cb_button_state_changed,
	                                 NULL);

	printf("Press key to exit\n");
	getchar();
	rgb_led_button_destroy(&rlb);
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
	return 0;
}

File: software/examples/rust/example_callback.rs
use std::{error::Error, io, thread};
use tinkerforge::{ipconnection::IpConnection, rgb_led_button_bricklet::*};

const HOST: &str = "127.0.0.1";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your RGB LED Button Bricklet

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection
    let rgb_led_button_bricklet = RGBLEDButtonBricklet::new(UID, &ipcon); // Create device object

    ipcon.connect(HOST, PORT).recv()??; // Connect to brickd
                                        // Don't use device before ipcon is connected

    //Create listener for button state changed events.
    let button_state_changed_listener = rgb_led_button_bricklet.get_button_state_changed_receiver();
    // Spawn thread to handle received events. This thread ends when the rgb_led_button_bricklet
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for event in button_state_changed_listener {
            if event == RGB_LED_BUTTON_BRICKLET_BUTTON_STATE_PRESSED {
                println!("State: Pressed");
            } else if event == RGB_LED_BUTTON_BRICKLET_BUTTON_STATE_RELEASED {
                println!("State: Released");
            }
        }
    });

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}
